docs/ocean/design_docs/vert-lagrang-remap/vert-lagrang-remap.rst

```diff
@@ -0,0 +1,524 @@
+   
+   \sum_{k=1}^{kmax} \phi_k^{t+1} h_k^{t+1} = \sum_{k=1}^{kmax} \phi_k^{lg} h_k^{lg}
+
+The vertical velocity across layer interfaces may be computed anytime after 
+regridding. It can be computed as 
+
+.. math::
+
+   w = - \nabla \cdot (h_k \mathbf{u}_k) - (h_k^{t+1} - h_k^t)/dt
+
+or
+
+.. math::
+
+   w = (h_k^{t+1} - h_k^lg)/dt
+
+The choice between the two is discussed in the Implementation section.
+
+
+Implementation
+--------------
+
+Implementation: Ability to perform vertical Lagrangian-remapping
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Date last modified: 2020/12/15
+
+Contributors: Darren Engwirda, Xylar Asay-Davis, Carolyn Begeman
+
+Namelist options:
+
+- To turn VLR on/off: 
+  :code:`ALE_vertical_grid, config_vert_lagrangian_remap = .true. or .false.`
+- *Something related to target grid, for now just z_initial*
+
+Namelist options specific to PPR are discussed later.
+
+Lagrangian step:
+
+The solution for prognostic variables in RK4 and split-explicit remains
+largely the same. The main difference is that the vertical velocity through 
+the top of layers is set to zero in the routine 
+:code:`ocn_vert_transport_velocity_top`. This is similar to what is done when 
+:code:`config_vert_coord_movement` is :code:`impermeable_interfaces`, except 
+rather than exit the routine, we proceed with computations needed for the z-star and
+z-tilde coordinate choices.
+
+*Other modifications to ocn_vert_transport_velocity_top are not yet determined.*
+We will likely need to bypass the :code:`ocn_ALE_thickness` call in 
+:code:`ocn_vert_transport_velocity_top` so that the adjustments of layer 
+thickness for SSH perturbations occur during the regridding step.
+
+
+Regridding step(s):
+
+#. :math:`z_k^1`, the depth of the top of the layer, is determined based on 
+   an analytical expression for the grid. 
+   The simplest case is constant z-levels, :math:`z_k^1 = z_k^{init}`.
+   Since :math:`z_k^1` can be a function of the ocean state (e.g., :math:`\rho` 
+   for isopycnal coordinates, regridding doesn't begin until after the solution 
+   for prognostic variables.
+#. Superimpose SSH perturbations according to one of the existing depth-
+   dependent functions, :math:`z_k^2 = z_k^1 + c(z) \: \eta`. As in 
+   :code:`ocn_ALE_thickness`, layer thicknesses are adjusted from the seafloor 
+   upwards. Ideally, there is a single function that is used for both ALE
+   implementations, with and without VLR.
+#. Apply conditioning steps outlined in the following section.
+#. Update :code:`layerThicknessEdge` from the updated :code:`layerThickness`
+   using routine :code:`ocn_diagnostic_solve_layerThicknessEdge`. Note that 
+   :code:`zTop, zMid` are updated later when :code:`ocn_diagnostic_solve` is 
+   called.
+
+All of the regridding steps will be performed from a separate module.
+This topic is further addressed in section Implementation: modularity.
+
+
+Remapping step:
```

Comment:
At least for purposes of this design, we should probably settle on a single term ("regridding" or "remapping").  My sense is that, in general, "remapping" is more general and includes the potential for unstructured meshes.  Either term is correct in this context but I might lean toward the more restrictive term "regridding".

Reply:
Reading further, I gather that "regridding" here refers to determining the new target grid, while "remapping" is putting the variables onto that target grid.  These uses of the terms would not be consistent with my understanding but may have been adopted in the ALE literature.

Reply:
@xylar, yes, I think the "regrid" + "remap" terms are consistent with e.g. HYCOM, MOM6 terminology, etc. See [Griffies et al](https://agupubs.onlinelibrary.wiley.com/doi/pdf/10.1029/2019MS001954), for example:
- "regrid": select new target grid levels.
- "remap": do the integral interpolation.

I'd personally be happy to use different terminology (others have also found the regrid / remap combination to be a little inscrutable in the past!) Perhaps: "grid selection" and "remap"?

Reply:
Yes, I'd prefer "grid selection" and "remap".

Reply:
I'm fine with changing the terminology as suggested. However, I think this does raise the question of where we want the `layerThickness` update to take place because in Griffies et al. this is at least conceptually associated with the regridding step. I think it does make sense to have the remapping module update both prognostic variables and layer thickness and just do grid selection in the regridding module. Sound good to you @xylar and @dengwirda ?

Reply:
Yes, I think grid selection would merely determine what the target vertical grid is.  This would probably be specified in terms of the target layer thickness on cells.  But this would not necessarily involve actually updating `layerThickness` or `layerThicknessEdge`.

I agree, then, that it would make sense to update `layerThickness` (and make the call to update `layerThicknessEdge`) within the remapping step.

So there would not be a "regridding module" but rather something like a "vertical grid" module, I would think.  We have something like that already in `mode_init` but it's nothing to use as a starting point.

Reply:
Yes, I think this makes sense: "grid selection" would determine the position of the new grid levels, and "remap" would update the prognostic variables, thicknesses, etc for the next time step.

Reply:
Keeping thickness and tracer remapping together also helps if you require tracer consistency, so another reason to do it this way.